devito/passes/clusters/buffering.py

```diff
@@ -94,206 +80,187 @@ def buffering(clusters, callback, sregistry, options, **kwargs):
-
-                guards[b.xd] = GuardBound(0, b.firstidx.f)
-
-            elif b.is_write and init_onwrite(b.function):
-                dims = b.buffer.dimensions
-                lhs = b.buffer.indexify()
-                rhs = 0
-
-            else:
-                # NOTE: a buffer must be initialized only if the buffered
-                # Function is read in at least one place or in the case of
-                # non-uniform SubDimensions, to avoid uninitialized values to
-                # be copied-back into the buffered Function
-                continue
-
-            expr = lower_exprs(Eq(lhs, rhs))
-            ispace = b.writeto
-            guards[pd] = GuardBound(dim.root.symbolic_min, dim.root.symbolic_max)
-            properties = {d: {AFFINE, PARALLEL} for d in ispace.itdims}
+        if not descriptors:
+            return clusters
 
-            init.append(Cluster(expr, ispace, guards=guards, properties=properties))
+        # Then we create the initializing Clusters when necessary
+        init = init_buffers(descriptors, self.options)
 
-        if not buffers:
-            return clusters
+        # Create all the necessary ModuloDimensions to step through the buffer
+        mds = make_mds(descriptors, prefix, self.sregistry)
+        subiters = as_mapper(mds.values(), lambda i: i.root)
 
         # Substitution rules to replace buffered Functions with buffers
+        # E.g., `usave[time+1, x+1, y+1] -> ub0[t1, x+1, y+1]`
         subs = {}
-        for b in buffers:
-            for a in b.accessv.accesses:
-                subs[a] = b.indexed[[b.index_mapper.get(i, i) for i in a.indices]]
+        for b, v in descriptors.items():
+            accesses = chain(*[c.scope[v.f] for c in v.clusters])
+            index_mapper = {i: mds[(v.xd, i)] for i in v.indices}
+            for a in accesses:
+                subs[a.access] = b.indexed[[index_mapper.get(i, i) for i in a]]
 
         processed = []
         for c in clusters:
             # If a buffer is read but never written, then we need to add
             # an Eq to step through the next slot
             # E.g., `ub[0, x] = usave[time+2, x]`
-            for b in buffers:
-                if not b.is_readonly:
+            for b, v in descriptors.items():
+                if not v.is_readonly:
                     continue
-                try:
-                    c.exprs.index(b.firstread)
-                except ValueError:
+                if c is not v.firstread:
                     continue
 
-                dims = b.function.dimensions
-                lhs = b.indexed[dims]._subs(b.dim, b.lastidx.b)
-                rhs = b.function[dims]._subs(b.dim, b.lastidx.f)
+                idxf = v.last_idx
+                idxb = mds[(v.xd, idxf)]
+
+                lhs = v.b.indexify()._subs(v.xd, idxb)
+                rhs = v.f.indexify()._subs(v.dim, idxf)
+
+                expr = Eq(lhs, rhs)
+                expr = lower_exprs(expr)
```

Comment:
Seems overkil, doesn't this just need `indexify`

Reply:
no, you have to shift the indices according to the halo as well

btw, this is nothing new, it's also in master